### PR TITLE
De-duplicate file-based logging implementations

### PR DESCRIPTION
**Pull request by basil**
### Problem

Logging in Winstone and Jenkins is still a bit of a mess. Winstone currently has an incomplete and not-yet-working implementation of a `--logfile` argument that logs _only_ Winstone logs to the given file. The fact that this is not-yet-working is technically a regression, since [pre-Jetty Winstone apparently had a working `--logfile` argument](https://winstone.sourceforge.net/)). The Jenkins main class avoids this not-yet-working implementation by intercepting the `--logfile` argument and handling it independently with its own implementation that redirects not just Winstone logs to the given file but _all_ logs.

Making matters worse, the Jenkins-specific version uses deprecated Java functionality that is slated for removal in later releases; namely, handling of the `ALRM` signal to facilitate log rotation. This turned out to be completely unnecessary as of recent releases: not only do most users of Jenkins run via Docker or `systemd` without a `--logfile` option and the need for log rotation support, but those who do rely on `--logfile` option and log rotation support no longer need signal handling support as of https://github.com/jenkinsci/packaging/pull/218, which switched the `logrotate` configuration of the official packages to using `copytruncate`, thereby obviating the need to handle `SIGALRM` in the application. So this is dead code, and it is a liability because it uses deprecated functionality.

### Solution

This PR simply de-duplicates the `--logfile` implementations by ripping out Jenkins' custom `--logfile` support and delegating to a new version of Winstone that now has a working implementation of `--logfile`, whose semantics in Winstone now extend to the whole application rather than just Winstone's own logs.

### Implementation

The implementation starts in Winstone, where we restore the generic Winstone `--logfile` argument (previously removed in favor of the Jenkins-specific one). Its implementation now redirects standard out and standard error to the given file using the default character set, just like the Jenkins-specific `--logfile` implementation did. While I was implementing this I found a lot of dead code that made it hard to understand what was going on, so I just deleted most of it. In particular, the `Logger` class had a lot of logic to store a bunch of streams… that were then never used. Eliminating the setting of these streams makes it more clear that this is just dead code. Both the old and new implementation rely on `ConsoleHandler` writing to standard error and the redirection of this stream to a given file if `--logfile` is provided. The `Logger` class also maintained its own custom levels for reasons that I could not understand, so I inlined these to remove an unnecessary layer of indirection in the calling code.

On the Jenkins side all that is needed was to remove the custom `--logfile` implementation and delegate to Winstone. This implied removing any usage of the logging subsystem from the Jenkins main class (formerly `extras-executable-war`), since doing so would initialize JUL to the wrong streams prior to Winstone getting a chance to parse the `--logfile` argument and redirect the standard streams. This is all explained in the code comments.

### Testing done

I tested this both with and without custom `--logfile` settings. With a custom `--logfile` option I verified that logs were properly sent to the given file rather than to systemd. Without the custom option I verified that the behavior remained the same as it is today.

The help now looks like this:

```
Options:
   --webroot                = folder where the WAR file is expanded into. Default is ${JENKINS_HOME}/war
   --pluginroot             = folder where the plugin archives are expanded into. Default is ${JENKINS_HOME}/plugins
                              (NOTE: this option does not change the directory where the plugin archives are stored)
   --extractedFilesFolder   = folder where extracted files are to be located. Default is the temp folder
   --enable-future-java     = allows running with Java versions which are not fully supported
   --paramsFromStdIn        = Read parameters from standard input (stdin)
   --version                = Print version to standard output (stdout) and exit
   --javaHome               = Override the JAVA_HOME variable
   --config                 = load configuration properties from here. Default is ./winstone.properties
   --prefix                 = add this prefix to all URLs (eg http://localhost:8080/prefix/resource). Default is none
   --commonLibFolder        = folder for additional jar files. Default is ./lib

   --extraLibFolder         = folder for additional jar files to add to Jetty classloader

   --logfile                = redirect log messages to this file
   --logThrowingLineNo      = show the line no that logged the message (slow). Default is false
   --logThrowingThread      = show the thread that logged the message. Default is false
   --debug                  = set the level of Winstone debug msgs (1-9). Default is 5 (INFO level)
[…]
```

### Future work

The `--logThrowingLineNo` and `--logThrowingThread` are pretty much useless; the former doesn't do anything at all and the latter just applies to Winstone logs, not all logs. I plan to eventually get rid of these options, but for now it is harmless to leave them in.

### Proposed changelog entries

Remove support for log rotation via `SIGALRM`.

### Proposed upgrade guidelines

Support for log rotation via `SIGALRM` has been removed. Users of the official Jenkins OS packages, which rely on the `copytruncate` option, are not affected. Users of the `--logfile` argument who have their own `logrotate` configuration should ensure they are using the `copytruncate` option rather than relying on `SIGALRM`.

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7256"></a>

### CODE REVIEW

File: pom.xml
Comment (basil) [inline pom.xml:101]: https://github.com/jenkinsci/winstone/pull/295

File: war/pom.xml
Comment (basil) [inline war/pom.xml:196]: No need for weird compiler options now that we are no longer using deprecated APIs.

File: war/src/main/java/executable/LogFileOutputStream.java
Comment (basil) [inline war/src/main/java/executable/LogFileOutputStream.java:76]: This is now handled by the call to `Files.newOutputStream` in Winstone's `Launcher.initLogger`.

File: war/src/main/java/executable/Main.java
Comment (basil) [inline war/src/main/java/executable/Main.java:380]: This is now handled by the call to `new PrintStream` in Winstone's `Launcher#initLogger`. Critically, we preserve compatibility by keeping the (implicit here, explicit in Winstone) default character set.

File: war/src/main/java/executable/Main.java
Comment (basil) [inline war/src/main/java/executable/Main.java:384]: The removal of the `--daemon` option here is unrelated code cleanup. I removed that option a long time ago but forgot to remove it from this little bit of code.

**Review by NotMyFault — APPROVED** [review]

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by StefanSpieker — APPROVED** [review]

File: war/src/main/java/executable/Main.java
Comment (basil) [reply war/src/main/java/executable/Main.java:385]: > we have some legacy systems still runnning with an old centOS 6

Jenkins does not support CentOS 6.

**Pull request merged**
